sdk/keyvault/azure-security-keyvault-administration/src/backup_client.cpp

```diff
@@ -0,0 +1,329 @@
+using namespace Azure::Core::Http;
+using namespace Azure::Core::Http::Policies;
+using namespace Azure::Core::Http::Policies::_internal;
+using namespace Azure::Core::Http::_internal;
+
+BackupClient::BackupClient(
+    std::string const& vaultUrl,
+    std::shared_ptr<Core::Credentials::TokenCredential const> credential,
+    BackupClientOptions options)
+    : m_vaultBaseUrl(vaultUrl), m_apiVersion(options.ApiVersion)
+{
+  std::vector<std::unique_ptr<HttpPolicy>> perRetrypolicies;
+  {
+    Azure::Core::Credentials::TokenRequestContext tokenContext;
+    tokenContext.Scopes = {_internal::UrlScope::GetScopeFromUrl(m_vaultBaseUrl)};
+
+    perRetrypolicies.emplace_back(
+        std::make_unique<_internal::KeyVaultChallengeBasedAuthenticationPolicy>(
+            credential, std::move(tokenContext)));
+  }
+  std::vector<std::unique_ptr<HttpPolicy>> perCallpolicies;
+
+  m_pipeline = std::make_shared<Azure::Core::Http::_internal::HttpPipeline>(
+      options,
+      _detail::KeyVaultServicePackageName,
+      _detail::PackageVersion::ToString(),
+      std::move(perRetrypolicies),
+      std::move(perCallpolicies));
+}
+
+Azure::Response<BackupOperation> BackupClient::FullBackup(
+    Azure::Core::Url const& blobContainerUrl,
+    SasTokenParameter const& sasToken,
+    Core::Context const& context)
+{
+  auto url = m_vaultBaseUrl;
+  url.AppendPath("backup");
+
+  url.SetQueryParameters({{"api-version", m_apiVersion}});
+
+  std::string jsonBody;
+  {
+    auto jsonRoot = Core::Json::_internal::json::object();
+
+    jsonRoot["storageResourceUri"] = blobContainerUrl.GetAbsoluteUrl();
+
+    if (sasToken.Token.HasValue())
+    {
+      jsonRoot["token"] = sasToken.Token.Value();
+    }
+
+    if (sasToken.UseManagedIdentity.HasValue())
+    {
+      jsonRoot["useManagedIdentity"] = sasToken.UseManagedIdentity.Value();
+    }
+
+    jsonBody = jsonRoot.dump();
+  }
+
+  Core::IO::MemoryBodyStream requestBody(
+      reinterpret_cast<std::uint8_t const*>(jsonBody.data()), jsonBody.length());
+
+  Core::Http::Request request(Core::Http::HttpMethod::Post, url, &requestBody);
+
+  request.SetHeader(HttpShared::ContentType, HttpShared::ApplicationJson);
+  request.SetHeader(HttpShared::Accept, HttpShared::ApplicationJson);
+  request.SetHeader("Content-Length", std::to_string(requestBody.Length()));
+
+  auto rawResponse = m_pipeline->Send(request, context);
+  auto const httpStatusCode = rawResponse->GetStatusCode();
+
+  if (httpStatusCode != Core::Http::HttpStatusCode::Accepted)
+  {
+    throw Core::RequestFailedException(rawResponse);
+  }
+
+  BackupOperationStatus response = DeserializeBackupOperationStatus(*rawResponse);
+  BackupOperation operation(std::make_shared<BackupClient>(*this), std::move(response), true);
+  return Response<BackupOperation>(std::move(operation), std::move(rawResponse));
+}
+
+Azure::Response<BackupOperationStatus> BackupClient::FullBackupStatus(
+    std::string const& jobId,
```

Comment:
Does the server really accept "null" as the job ID?

Reply:
nope, since the jobid endsup as part of the url hit it will return a 404

Reply:
If the server is always going to reject the request, this is the kind of situation where I think it's appropriate to throw an exception in the client if the `jobId` parameter is empty.

As a developer, I would probably prefer to get a runtime exception saying "the jobId parameter cannot be empty" rather than a 404 exception.

And that also allows you to skip over this special case.